boot/freeldr/bootsect/ntfs.S

```diff
@@ -0,0 +1,829 @@
+        jmp   NodeCheck
+
+    ExitIndexTree:
+    pop   di
+    pop   si
+    mov   sp, bp
+    pop   bp
+    ret
+
+// [ES:SI] - the source file name
+// [DS:DI] - the destination file name
+// CX - compare length
+CompareStr:
+    push bp
+    mov  bp, sp
+    push bx
+    push si
+    push di
+    movzx cx, cl
+    CmpLoop:
+        mov ax, word ptr es:[si]
+        cmp ax, 'a'
+        jl NoUpper
+        cmp ax, 'z'
+        jg NoUpper
+        sub ax, HEX(20)
+        NoUpper:
+        mov bx, word ptr ds:[di]
+        sub bx, ax
+        test bx, bx
+        jnz  CompareFail
+        add  si, 2
+        add  di, 2
+        dec cx
+        jnz CmpLoop
+    CompareFail:
+    mov ax, bx
+    pop di
+    pop si
+    pop bx
+    mov sp, bp
+    pop bp
+    ret
+
+// Reads a NTFS cluster
+// AX     : cluster number to read
+// ES:BX  : address to read
+ReadCluster:
+    push  edx
+    push  ecx
+    mov   cl, byte ptr [SectsPerCluster]   // Convert sectors to clusters
+    movzx ecx, cx
+    mul   ecx
+    add   eax, dword ptr [HiddenSectors]   // Add offset from the disk beginning
+    call  ReadSectors
+    pop   ecx
+    pop   edx
+    ret
+
+// Reads a MFT entry
+// AX     : MFT inode
+// ES:BX  : address to read
+ReadInode:
+    push  edx
+    push  ecx
+    push  si
+    push  di
+
+    mov   cx, word ptr [SectsPerMFT]       // Get the correct number of sectors for the FILE entry
+    mul   cx
+    movzx eax, ax
+    mov   ecx, dword ptr [MFTStartSector]  // Add it to the start of the MFT
+    add   eax, ecx
+    add   eax, dword ptr [HiddenSectors]   // Add disk offset
+    mov   cx, word ptr [SectsPerMFT]
+    movzx ecx, cx
+    call  ReadSectors
+    
+    cmp   dword ptr es:[bx], HEX(454c4946) // Ensure we get a valid FILE record
+    jnz   near ptr PrintFileSystemError
+
+    call  ApplyFixups
+
+    pop   di
+    pop   si
+    pop   ecx
+    pop   edx
+    ret
+
+#define UpdateSequenceOffset 4
+#define UpdateSequenceLen 6
+
+// Apply fixups to INDX and FILE records
+// ES:BX - record to fixup
+ApplyFixups:
+    push  si
+    push  di
+    mov   si, bx
+    add   si, word ptr es:[bx + UpdateSequenceOffset]
+    xor   cx, cx
+    inc   cx
+    FixupLoop:
+        cmp  cx, word ptr es:[bx + UpdateSequenceLen]
+        jz  EndFixupLoop
+        mov   si, bx
+        add   si, word ptr es:[bx + UpdateSequenceOffset]
+        mov   ax, word ptr es:[si]     // Get first fixup value
+
+        mov   di, cx
+        shl   di, 9
+        add   di, bx
+        sub   di, 2
+        cmp   ax, word ptr es:[di]     // Check fixup value
+        jnz   PrintFileSystemError     // Fixup is corrupted, so print error
+        inc   cx
+        add   si, cx
+        mov   ax, word ptr es:[si]     // Apply fixup
+        mov   word ptr es:[di], ax
+
+        jmp   FixupLoop
+    EndFixupLoop:
+
+    pop di
+    pop si
+    ret
+
+// Reads a non-resident attribute
+// ES:AX : Address of the data runs
+// FS:BX : Address to write to
```

Comment:
This looks a bit unusual. Why not `DS:AX` and `ES:BX` (to fit the `ReadCluster`)?

Reply:
Because I didn't want to touch DS, since it was useful to print debug messages on that part